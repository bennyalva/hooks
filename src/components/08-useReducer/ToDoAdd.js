import React from 'react'
import { useForm } from '../../hooks/useForm';

export const ToDoAdd = ({todoAdd}) => {
    const [{description}, handalerInputChange, reset] = useForm({description: ''});
    const handlerSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length <= 1 ) {
            return;
        }

        const newToDo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }
        todoAdd({type: 'add', payload: newToDo});
        reset();
     }
    return (
        <>
         <h4>Agregar ToDo</h4>
                <hr />
                <form onSubmit={handlerSubmit}>
                    <input 
                            type="text"
                            name="description"
                            placeholder="Learn .."
                            autoComplete="off"
                            className="form-control"
                            onChange={handalerInputChange}
                            value={description}
                            />
                            <button
                                    type="submit"
                                    className="btn btn-outline-primary mt-1 btn-block">
                                    Agregar
                            </button>
                </form>   
        </>
    )
}
