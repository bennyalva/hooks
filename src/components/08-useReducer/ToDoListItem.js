import React from 'react'

export const ToDoListItem = ({toDO, i, handlerDelete, handlerToggle}) => {
    return (
        <li
                    key={toDO.id}
                    className="list-group-item">
                    <p className={`${ toDO.done && 'complete'}`}
                        onClick={ () => handlerToggle(toDO.id)}>{i +1}. {toDO.desc}</p>
                    <button
                        onClick={() => handlerDelete(toDO.id)}
                        className="btn btn-danger">
                            Borrar
                        </button>
                </li>
    )
}
