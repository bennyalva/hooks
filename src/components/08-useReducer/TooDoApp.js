import React, { useReducer } from 'react'
import './styles.css';
import { toDoReducer } from './toDoReducer';
const initialState = [
    {
        id: new Date().getTime(),
        desc: 'Que paso rey',
        done: false
    }
]
export const TooDoApp = () => {

    const [toDos] = useReducer(toDoReducer, initialState);
    console.log(toDos)
    return (
        <div>
            <h1>ToDoApp ({toDos.length})</h1>
            <hr />
          <div className="row">
            <div className="col-7">
                <ul className="list-group list-group-flush">
                    {
                        toDos.map((toDO, i) => (
                            <li
                                key={toDO.id}
                                className="list-group-item">
                                <p className="text-center">{i +1}. {toDO.desc}</p>
                                <button
                                    className="btn btn-danger">
                                        Borrar
                                    </button>
                            </li>
                        ))
                    }
                </ul>
            </div>
            <div className="col-5">
                <h4>Agregar ToDo</h4>
                <hr />
                <form>
                    <input 
                            type="text"
                            name="description"
                            placeholder="Learn .."
                            autoComplete="off"
                            className="form-control"
                            />
                            <button
                                    className="btn btn-outline-primary mt-1 btn-block">
                                    Agregar
                            </button>
                </form>
            </div>
          </div>
        </div>
    )
}
