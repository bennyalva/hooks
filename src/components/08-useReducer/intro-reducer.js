const initialState = [
    {
        id: 1,
        toDo: 'Comprar pan',
        done: false
    }
]
const toDoReducer = (state = initialState, action) => {

    if (action?.type === 'agregar') {
        return [...state, action.payload];
    }
    return state;
}

let toDos = toDoReducer();

const newToDo = {
    id: 2,
    toDo: 'Comprar leche',
    done: false
}

const action = {
    type: 'agregar',
    payload: newToDo
}

toDos = toDoReducer(toDos, action)

console.log(toDos)