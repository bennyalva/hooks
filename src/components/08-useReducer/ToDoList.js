import React from 'react'
import { ToDoListItem } from './ToDoListItem'

export const ToDoList = ({toDos, handlerDelete, handlerToggle}) => {
    return (
        <ul className="list-group list-group-flush">
        {
            toDos.map((toDO, i) => (
                <ToDoListItem
                            key={toDO.id} 
                            toDO={toDO} 
                            i={i} 
                            handlerDelete={handlerDelete} 
                            handlerToggle={handlerToggle}/>
                // <li
                //     key={toDO.id}
                //     className="list-group-item">
                //     <p className={`${ toDO.done && 'complete'}`}
                //         onClick={ () => handlerToggle(toDO.id)}>{i +1}. {toDO.desc}</p>
                //     <button
                //         onClick={() => handlerDelete(toDO.id)}
                //         className="btn btn-danger">
                //             Borrar
                //         </button>
                // </li>
            ))
        }
    </ul>
    )
}
