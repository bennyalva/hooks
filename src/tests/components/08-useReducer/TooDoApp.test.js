import React from 'react';
import { shallow, mount } from 'enzyme';
import { TooDoApp } from '../../../components/08-useReducer/TooDoApp';
import { demoToDos } from '../../fixtures/demoToDos';
import { act } from '@testing-library/react';
describe('tests on <TooDoApp />', () => {
    const wrapper = shallow(<TooDoApp />); // da información general del componente
    Storage.prototype.setItem = jest.fn(() => {}); // to simulate stogare
    test('should show correctly', () => {
       expect(wrapper).toMatchSnapshot(); 
    });
    test('should add a toDo', () => {
    //    const wrapper = mount(<TooDoApp />); // da más información del componente y sus hijos
      act(() => {
           // wrapper.find('ToDoAdd').prop('handlerAddToDo');
            // console.log('compo: ', wrapper.find('ToDoAdd'))
        });
      //   expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });
    test('should delete a toDo', () => {
       //  wrapper.find('ToDoAdd').prop('handlerAddToDo')(demoToDos[0]);
        // wrapper.find('ToDoList').prop('handlerDelete')(demoToDos[0].id);
        // check because no work
        // expect(wrapper.find('h1').text().trim()).toBe('ToDoApp ( 0 )')

    });
    
    
})
