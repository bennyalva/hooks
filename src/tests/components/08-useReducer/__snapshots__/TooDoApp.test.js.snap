// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <TooDoApp /> should show correctly 1`] = `
<div>
  <h1>
    ToDoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <ToDoList
        handlerDelete={[Function]}
        handlerToggle={[Function]}
        toDos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <ToDoAdd
        todoAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
